liquidity-bot/src/lib.rs

```diff
@@ -2,10 +2,1074 @@
+            .ceil() // We round up in case the conversion results in a fractional amount.
+            .to_u64()
+            .ok_or_else(|| {
+                Error::DecimalConversion(format!(
+                    "Could not convert {} times {} to u64",
+                    matched_tx_out.amount.value, swap_rate
+                ))
+            })?;
+
+        log::info!(
+            self.logger,
+            "Creating SCI for TxOut with amount {}: wanting {} counter tokens (token id {})",
+            matched_tx_out.amount.value,
+            counter_amount,
+            counter_token_id,
+        );
+
+        // Construct a ring. The first step is to choose tx out indices.
+        let mut rng = rand::thread_rng();
+        let mut sampled_indices = HashSet::new();
+
+        const RING_SIZE: usize = 11;
+        let num_txos = self.ledger_db.num_txos()?;
+        while sampled_indices.len() < RING_SIZE {
+            let index = rng.gen_range(0..num_txos);
+            sampled_indices.insert(index);
+        }
+        let mut sampled_indices_vec: Vec<u64> = sampled_indices.into_iter().collect();
+
+        // Ensure our TxOut is in the ring. Always using index 0 is safe since
+        // InputCredentials::new sorts the ring.
+        let our_tx_out_index = self
+            .ledger_db
+            .get_tx_out_index_by_public_key(&matched_tx_out.tx_out.public_key)?;
+        sampled_indices_vec[0] = our_tx_out_index;
+
+        // Get the actual TxOuts
+        let tx_outs = sampled_indices_vec
+            .iter()
+            .map(|idx| self.ledger_db.get_tx_out_by_index(*idx))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // Get proofs for all those indexes.
+        let proofs = self
+            .ledger_db
+            .get_tx_out_proof_of_memberships(&sampled_indices_vec)?;
+
+        // Create our InputCredentials
+        let input_credentials = InputCredentials::new(
+            tx_outs,
+            proofs,
+            0,
+            OneTimeKeyDeriveData::SubaddressIndex(matched_tx_out.subaddress_index),
+            *self.account_key.view_private_key(),
+        )?;
+
+        // Build the SCI
+        let block_version = BlockVersion::MAX;
+        let fog_resolver = FogResolver::default(); // TODO
+
+        let mut builder = SignedContingentInputBuilder::new(
+            block_version,
+            input_credentials,
+            fog_resolver,
+            EmptyMemoBuilder::default(), // TODO
+        )?;
+        builder.add_partial_fill_output(
+            Amount::new(counter_amount, *counter_token_id),
+            &self.account_key.default_subaddress(),
+            &mut rng,
+        )?;
+        builder.add_partial_fill_change_output(
+            matched_tx_out.amount,
+            &ReservedSubaddresses::from(&self.account_key),
+            &mut rng,
+        )?;
+        let sci = builder.build(&LocalRingSigner::from(&self.account_key), &mut rng)?;
```

Comment:
it looks that we are passing real proofs to the SCI builder, so i think the tx_out_global_indices is set.

But maybe we should clear the proofs from the SCI that is created?

Maybe that should be something the SCI builder does automatically

Reply:
i had an issue in mobilecoind at one point where the tx_out_global_indices were all zeroes, so i'm just trying to think through what is happening here

Reply:
I'm fairly certain the indices are not zero since aside from looking at the code, I ran this locally and its SCIs got accepted by a local DEQS (that verifies the indices). 

It looks like the builder is already clearing the proofs: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/builder/src/signed_contingent_input_builder.rs#L510
But maybe the DEQS should ensure that they are in fact clear? otherwise people can spam large SCIs